File: pc/McController.Server/Net/Messages.cs
namespace McController.Server.Net;

/// <summary>
/// Decoded form of a TCP control packet. Discriminated by record type
/// so callers can pattern-match.
/// </summary>
public abstract record ControlMessage;

public sealed record HelloMsg(byte ProtoVer, uint ClientId, bool WantsUdp) : ControlMessage;
public sealed record JoystickMsg(float X, float Y) : ControlMessage;
public sealed record LookDeltaTcpMsg(uint Seq, short Dx, short Dy) : ControlMessage;
public sealed record ButtonMsg(byte ButtonId, bool Down) : ControlMessage;
public sealed record PingMsg(uint Seq) : ControlMessage;

/// <summary>Unrecognized type byte — caller can log and ignore.</summary>
public sealed record UnknownMsg(byte Type, int PayloadLength) : ControlMessage;

/// <summary>UDP camera datagram. Carried on its own (non-TCP) channel.</summary>
public sealed record LookDeltaUdpMsg(uint Seq, short Dx, short Dy);

File: mac/McController/Core/Net/Messages.swift
import Foundation

/// Decoded form of a TCP control packet. Mirrors `Messages.cs` on the PC side.
enum ControlMessage: Equatable {
    case hello(protoVer: UInt8, clientId: UInt32, wantsUdp: Bool)
    case joystick(x: Float, y: Float)
    case lookDeltaTcp(seq: UInt32, dx: Int16, dy: Int16)
    case button(id: UInt8, down: Bool)
    case ping(seq: UInt32)
    case probe
    case probeAck(status: UInt8)
    case unknown(type: UInt8, payloadLength: Int)
}

/// UDP camera datagram carried on its own (non-TCP) channel.
struct LookDeltaUdpMsg: Equatable {
    let seq: UInt32
    let dx: Int16
    let dy: Int16
}
